MAVProxy/modules/mavproxy_magical/__init__.py

```diff
@@ -0,0 +1,194 @@
+# Copyright (C) 2016  Intel Corporation. All rights reserved.
+#
+# This file is free software: you can redistribute it and/or modify it
+# under the terms of the GNU General Public License as published by the
+# Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This file is distributed in the hope that it will be useful, but
+# WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+# See the GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License along
+# with this program.  If not, see <http://www.gnu.org/licenses/>.
+"""
+Compass calibration helper. This is adapted from Ardupilot's local MAVProxy
+module magcal_graph. Some people read it "magical_graph" at the time of
+release, thus the silly reason for the name of this module.
+"""
+import multiprocessing
+
+from pymavlink import mavutil
+
+from MAVProxy.modules.lib import mp_module, mp_util
+
+class MagicalModule(mp_module.MPModule):
+    def __init__(self, mpstate):
+        super(MagicalModule, self).__init__(mpstate, 'magical')
+        self.add_command(
+            'magical_ui',
+            self.cmd_magical_ui,
+            'open the GUI for compass calibration',
+        )
+
+        self.mpstate = mpstate
+        self.parent_pipe, self.child_pipe = multiprocessing.Pipe()
+        self.ui_process = None
+        self.progress_msgs = {}
+        self.report_msgs = {}
+        self.attitude_msg = None
+        self.raw_imu_msg = None
+        self.running = False
+        self.last_ui_msgs = {}
+
+    def start_ui(self):
+        if self.ui_is_active():
+            return
+        if self.ui_process:
+            self.ui_process.join()
+
+        self.ui_process = multiprocessing.Process(target=self.ui_task)
+        self.update_ui()
+        self.ui_process.start()
+
+    def stop_ui(self):
+        if not self.ui_is_active():
+            return
+
+        self.parent_pipe.send(dict(name='close'))
+        self.ui_process.join()
```

Comment:
Perhaps we could have something a little more definite in terms of killing the UI process in case of it hanging.  Timeout the join and kill it, perhaps?